README.md
## About

`lemmagen3` is a Python2/3 wrapper for the [Lemmagen lemmatizer](http://lemmatise.ijs.si/).

It is different from other Lemmagen wrappers like [this one on PyPi](https://pypi.org/project/Lemmagen/) because it offers a clean, fast OO interface built with the excellent [pybind11 library](https://github.com/pybind/pybind11) and supports an additional language (Croatian).

Model for Slovene, Croatian and Serbian are significantly updated and make use of frequency data to prefer most frequent lemmas, e.g., for Slovene: `je->biti` instead of `je->jesti` and `od->od` instead of `od->oda`.

In total, 19 languages are supported:

1.  Bulgarian: `bg`
2.  Croatian: `hr`
3.  Czech: `cs`
4.  English: `en`
5.  Estonian: `et`
6.  Farsi/Persian: `fa`      
7.  French: `fr`  
8.  German: `de`
9.  Hungarian: `hu`
10. Italian: `it`
11. Macedonian: `mk`        
12. Polish: `pl`     
13. Romanian: `ro`            
14. Russian: `ru`             
15. Serbian: `sr`
16. Slovak: `sk`
17. Slovene: `sl`          
18. Spanish: `es`
19. Ukrainian: `uk`


## Installation and requirements


```sh
pip install lemmagen3
```
will install the module for `python 2.7+` and `python 3.5+`.

If you want to build it from source, you will need `pybind11` and a C++ compiler.
Then run

```sh
python setup.py install
```
which will compile and install the module.

**Note:**  If you use `python 3.5.0` or `3.5.1` you will likely get the error shown below. This is a known bug in these two versions so please consider upgrading your Python.

```sh
ImportError: ..._lemmagen.cpython-35m-x86_64-linux-gnu.so: undefined symbol: _PyThreadState_UncheckedGet
```


## How to use

The following snippet illustrates how to use `lemmagen3`.

```python
from lemmagen3 import Lemmatizer

# first, list all supported languages
print(Lemmatizer.list_supported_languages())

# then, create few lemmatizer objects using ISO 639-1 language codes
# (English, Slovene and Russian)

lem_en = Lemmatizer('en')
lem_sl = Lemmatizer('sl')
lem_ru = Lemmatizer('ru')

# now lemmatize the word "cats" in all three languages
print(lem_en.lemmatize('cats'))
print(lem_sl.lemmatize('mačke'))
print(lem_ru.lemmatize('коты'))

# you can also change the language for an existing Lemmatizer object
# lem_en will now become a French lemmatizer:
lem_en.load_language('fr')

# finally, you can also load your own Lemmagen model
my_lem = Lemmatizer()
my_lem.load_model('/path/to/my/model')
```

Note that the function `lemmatize` accepts single string tokens and **does not split the input string**! If you want to lemmatize a chunk of text you will have to tokenize it first, e.g.:

```python
sentence = 'cats hate dogs'
tokens = sentence.split()
sentence_lemmatized = ' '.join([lem_en.lemmatize(token) for token in tokens])
```

Note also that `lemmagen3` operates on unicode encoded strings so if you use python 2 make sure that your input string is encoded as unicode.


## License

Please note that this repository contains code and binary models compiled and built from different sources which are under different licenses:

1.  C++ files and headers come from [Lemmagen](http://lemmatise.ijs.si/) and are modified and adapted to work as a Python module (**LGPL**)
2.  Binary models are built from Multext and [Multext-east](http://nl.ijs.si/ME/V4/) sources:
    *   Language resources used to build Farsi/Persian, Macedonian, Polish, and Russian models are for **non-commercial use only**.
    *   Language resource for other supported languages are released under **CC BY-SA 4.0**.

The rest of the code in this repository was created by the author and is licensed under the **MIT license**.


## Authors

-   `lemmagen3` is developed by Vid Podpečan (vid.podpecan@ijs.si).
-   The Lemmagen lemmatizer was developed by [Matjaž Juršič](http://lemmatise.ijs.si/Home/Contact).


CHANGELOG
Version 3.3.1
-------------
Models for Slovene (sl), Croatian (hr) and Serbian (sr) are improved.

Version 3.2.2
-------------
TravisCI is now used for building wheels for all platforms (Linux, Mac, Windows) and the code was patched for manylinux2010 requirements.

Version 3.2.1
-------------
This is the first public release which follows two internal unreleased versions.


test/test.py
from lemmagen3 import Lemmatizer

print(Lemmatizer.list_supported_languages())

a = Lemmatizer('en')
word = 'cats'
print('{}->{}'.format(word, a.lemmatize(word)))

b = Lemmatizer('sl')
word = 'ljudje'
print('{}->{}'.format(word, b.lemmatize(word)))


src/lemmagen3/__init__.py
class Lemmatizer:
    '''
    This is a thin convenience wrapper aroung RdrLemmatizer which allows to use
    language codes for loading lemmatizer models instead of model file names.
    '''
    def __init__(self, lang=None):
        # import here to avoid namespace pollution
        from . _lemmagen import RdrLemmatizer
        self.__lemmatizer = RdrLemmatizer()
        if lang is not None:
            self.load_language(lang)

    def lemmatize(self, word):
        '''
        Lemmatizes given word.

        Parameters
        ----------
        word : str
            Input word.

        Returns
        -------
        str
            Lemmatized form of the input word.
        '''
        return self.__lemmatizer.lemmatize(word)

    def load_model(self, model):
        '''
        Loads a binary model.

        Parameters
        ----------
        model : str
            Path to the model file.
        '''
        self.__lemmatizer.load_model(model)

    def load_language(self, lang):
        '''
        Loads a model for a given supported language.

        Parameters
        ----------
        lang : str
            Language code in ISO 639-1 (two-letter code).
        '''
        from os.path import join, dirname, abspath, exists
        mpath = join(dirname(abspath(__file__)), 'models', '{}.bin'.format(lang))
        if not exists(mpath):
            raise ValueError('Language "{}" is not supported'.format(lang))
        self.__lemmatizer.load_model(mpath)

    @staticmethod
    def list_supported_languages():
        '''
        Returns a list of supported languages (names of present model files).
        '''
        from os.path import join, dirname, abspath
        from os import listdir
        return sorted([mfname.replace('.bin', '') for mfname in listdir(join(dirname(abspath(__file__)), 'models'))])


src/lemmagen3/cpp/definitions.h
/******************************************************************************
This file is part of the lemmagen library. It gives support for lemmatization.
Copyright (C) 2006-2007 Matjaz Jursic <matjaz@gmail.com>

The lemmagen library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
******************************************************************************/
//#pragma once

#define AllInOneFile

//-------------------------------------------------------------------------------------------
//includes
#include <iostream>
#include <fstream>
#include <iomanip>
#include <sstream>

#ifndef AllInOneFile
	#include "RdrLemmData.h"
#endif

//-------------------------------------------------------------------------------------------
//if data is not loaded from separate file define empty structure
#ifndef RrdLemmData
	#define RrdLemmData
	#define DATA_LEN 8
	#define DATA_TBL {0x0000000000000000}
#endif

//-------------------------------------------------------------------------------------------
//typedefs
// typedef unsigned char byte;
typedef unsigned short word;
typedef unsigned int dword;
typedef unsigned long long qword;

//-------------------------------------------------------------------------------------------
//const variables that algorithm depends on
#define AddrLen 4
#define FlagLen 1
#define ModLen 1
#define LenSpecLen 1
#define CharLen 1
#define DataStart 0

#define RoundDataLen 8

#define BitDefault	0x00
#define BitAddChar	0x01
#define BitInternal	0x02
#define BitEntireWr	0x04

#define TypeRule		(BitDefault)
#define TypeRuleEw		(BitDefault | BitEntireWr)
#define TypeLeafAC		(BitDefault | BitAddChar)
#define TypeLeafACEw	(BitDefault | BitAddChar | BitEntireWr)
#define TypeIntr		(BitDefault | BitInternal)
#define TypeIntrAC		(BitDefault | BitAddChar | BitInternal)

//-------------------------------------------------------------------------------------------
//main data structure and it's length
static dword iDataLenStatic = DATA_LEN;
static qword abDataStatic[] = DATA_TBL;

//-------------------------------------------------------------------------------------------
//helper macros for nicer code and faster execution
#if AddrLen == 3
	#define GETDWORD(type, wVar, wAddr) \
				type wVar = *((dword*) &abData[wAddr]) & 0x00FFFFF
#else
	#define GETDWORD(type, wVar, wAddr) \
				type wVar = *((dword *) &abData[wAddr])
#endif

#define GETBYTEMOVE(type, bByte, iSize) \
			type bByte = abData[iAddr]; \
			iAddr += iSize

#define GETDWORDMOVE(type, wVar, iSize) \
			GETDWORD(type, wVar, iAddr);  \
			iAddr += iSize

#define GETSTRINGMOVE(type, acString, iSize) \
			type acString = new char[iSize+1]; \
			strncpy(acString, (char*) &abData[iAddr], iSize); \
			acString[iSize] = NULL; \
			iAddr += iSize

//-------------------------------------------------------------------------------------------
using namespace std;

class RdrLemmatizer{
public:
	unsigned char *abData;
	int iDataLen;

public:
	RdrLemmatizer(const char *acFileName);
	RdrLemmatizer();
	~RdrLemmatizer();

	int SizeOfTree() const;

	char *Lemmatize(const char *acWord) const;

	void ToStringHex(ostream &os) const;
	void ToString(ostream &os = cout, dword iStartAddr = DataStart, int iDepth = 0,
		char *acParSufx = "", char *acParDev = "", char cNewChar=NULL) const;

	void LoadBinary(const char *acFileName);
	void LoadBinaryStream(istream &is);

};


src/lemmagen3/cpp/lemmagen.cpp
/******************************************************************************
This file is part of the lemmagen library. It gives support for lemmatization.
Copyright (C) 2006-2007 Matjaz Jursic <matjaz@gmail.com>
Minor modifications for a working Python extension by Vid Podpecan <vid.podpecan@ijs.si>

The lemmagen library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
******************************************************************************/

#include "definitions.h"
#include <pybind11/pybind11.h>

namespace py = pybind11;



PYBIND11_MODULE(_lemmagen, m) {
    py::class_<RdrLemmatizer>(m, "RdrLemmatizer")
        .def(py::init<const char *>(), "Creates a RdrLemmatizer and loads the given model")
 		.def(py::init(), "Creates an empty RdrLemmatizer")
        .def("load_model", &RdrLemmatizer::LoadBinary, "Loads a binary model")
        .def("lemmatize", &RdrLemmatizer::Lemmatize, "Lemmatizes given word");
}


//-------------------------------------------------------------------------------------------
//constructors
RdrLemmatizer::RdrLemmatizer(const char *acFileName) {
	LoadBinary(acFileName);
}

RdrLemmatizer::RdrLemmatizer() {
	this->abData = (unsigned char*) abDataStatic;
	this->iDataLen = iDataLenStatic;
}

//-------------------------------------------------------------------------------------------
//destructor
RdrLemmatizer::~RdrLemmatizer() {
	if (this->abData != (unsigned char*) abDataStatic)
		delete[] abData;
}

//-------------------------------------------------------------------------------------------
//returns size of a tree (number of bytes)
int RdrLemmatizer::SizeOfTree() const{
	return iDataLen - DataStart;
}

//-------------------------------------------------------------------------------------------
//lematizes word according to this data
char *RdrLemmatizer::Lemmatize(const char *acWord) const{
	unsigned char bWordLen = strlen(acWord);

	dword iAddr = DataStart;
	dword iParentAddr = DataStart;
	dword iTmpAddr;
	char bLookChar = bWordLen;
	unsigned char bType = abData[iAddr];

    if (this->abData == (unsigned char*) abDataStatic) {
        throw std::runtime_error("Cannot lemmatize: please load a model first.");
    }

	while(true) {
		iTmpAddr = iAddr+FlagLen+AddrLen;

		//check if additional characters match
		if ((bType & BitAddChar) == BitAddChar) {
			unsigned char bNewSufxLen = abData[iTmpAddr];
			iTmpAddr += LenSpecLen;

			bLookChar -= bNewSufxLen;

			//test additional chars if ok
			if (bLookChar>=0)
				do bNewSufxLen--;
				while (bNewSufxLen!=255 && abData[iTmpAddr+bNewSufxLen] == (unsigned char) acWord[bLookChar+bNewSufxLen]);

			//wrong node, take parents rule
			if (bNewSufxLen!=255) {	iAddr = iParentAddr; break; }

			//right node, but we are at the end (there will be no new loop) finish by returning this rule
			if ((bType & ~BitEntireWr) == TypeLeafAC) break;

			//right node and we need to go on with subnodes (it si probably type TypeIntrAC )
			//set iTmpAddr to start of hashtable
			iTmpAddr += abData[iTmpAddr-LenSpecLen];
		}

		//move lookup char back
		bLookChar--;
		//check if we are still inside the word (bLookChar==0 when at the begining of word)
		if (bLookChar<0) {
			//this means that we are just one character in front of the word so we must look for entireword entries
			if((bType & BitInternal) == BitInternal) {
				//go to the hashtable position 0(NULL) and look idf address is not NULL
				iTmpAddr += ModLen;
				unsigned char bChar = abData[iTmpAddr];
				GETDWORD(,iTmpAddr,iTmpAddr+CharLen);
				if (bChar == NULL && iTmpAddr!=NULL) {
					//we have a candidate for entireword, redirect addresses
					iParentAddr = iAddr;
					iAddr = iTmpAddr;
					bType = abData[iAddr];
					//increase lookchar (because we actualy eat one character)
					bLookChar++;
				}
			}
			break;
		}

		//find best node in hash table
		if((bType & BitInternal) == BitInternal) {
			unsigned char bMod = abData[iTmpAddr];
			unsigned char bChar = acWord[bLookChar];

			iTmpAddr += ModLen + (bChar%bMod)*(AddrLen+CharLen);

			iTmpAddr = abData[iTmpAddr] == bChar ? iTmpAddr + CharLen : iAddr + FlagLen;

			iParentAddr = iAddr;
			GETDWORD(,iAddr, iTmpAddr);
			bType = abData[iAddr];

			if ((bType & ~BitEntireWr) == TypeRule) break;
		}
	}
	//if this is entire-word node, and we are not at the begining of word it's wrong node - take parents
	if((bType & BitEntireWr) == BitEntireWr && bLookChar!=0) {
		iAddr = iParentAddr;
		bType = abData[iAddr];
	}

	//search ended before we came to te node of type rule but current node is OK so find it's rule node
	if((bType & ~BitEntireWr) != TypeRule)  GETDWORD( ,iAddr, iAddr+FlagLen);

	//we have (100%) node of type rule for lemmatization - now it's straight forward to lemmatize
	//read out rule
	iTmpAddr = iAddr + FlagLen;
	unsigned char iFromLen = abData[iTmpAddr];
	iTmpAddr += LenSpecLen;
	unsigned char iToLen = abData[iTmpAddr];
	iTmpAddr += LenSpecLen;

	//prepare output buffer
	unsigned char iStemLen = bWordLen - iFromLen;
	char *acReturn = new char[iStemLen + iToLen + 1];

	//do actual lematiration using given rule
	memcpy(acReturn, acWord, iStemLen);
	memcpy(&acReturn[iStemLen], &abData[iTmpAddr], iToLen);
	acReturn[iStemLen + iToLen] = NULL;

	return acReturn;
}

//-------------------------------------------------------------------------------------------
//returns string representation of this node and all its subnodes
void RdrLemmatizer::ToString(ostream &os, dword iStartAddr, int iDepth, char *acParSufx, char *acParDev, char cNewChar) const {
	int iAddr = iStartAddr;
	dword *iSubs = NULL;
	unsigned char *bSubs = NULL;
	int iSubsNum = 0;
	char *acSufx = NULL;
	char *acSufxDev = NULL;

	//node type
	GETBYTEMOVE(int, iType, FlagLen);

	char *acTypeName;
	switch (iType) {
		case TypeRule:     acTypeName="RULE"; break;
		case TypeRuleEw:   acTypeName="RULE(entireword)"; break;
		case TypeLeafAC:   acTypeName="LEAF"; break;
		case TypeLeafACEw: acTypeName="LEAF(entireword)"; break;
		case TypeIntr:     acTypeName="INTER-SHORT"; break;
		case TypeIntrAC:   acTypeName="INTER-LONG"; break;
	}

	os << setfill('\t') << setw(iDepth) << "" << "" << acTypeName << ":[Addr:" << iStartAddr << "]";

	//we have rule to display
	if((iType & ~BitEntireWr)==TypeRule) {
		GETBYTEMOVE(int, iFromLen, LenSpecLen);
		GETBYTEMOVE(int, iToLen, LenSpecLen);
		GETSTRINGMOVE(char*, acTo, iToLen);
		os << "[From:" << iFromLen << "][To:" << iToLen << ",\"" << acTo << "\"]";
	}

	//we have node to display
	else {
		//eat up rule addres
		GETDWORDMOVE(dword, iRuleAddress, AddrLen);

		//eat up characters
		int iNewSufxLen = 0;
		char* acNewSuffix;
		if ((iType & BitAddChar) == BitAddChar) {
			GETBYTEMOVE(, iNewSufxLen, LenSpecLen);
			GETSTRINGMOVE(, acNewSuffix, iNewSufxLen);
		}

		//create and display sufixes
		if (cNewChar != NULL) {
			int iSufxLen = 1 + iNewSufxLen + strlen(acParSufx) + 1;
			int iSufxDevLen = 2 + iNewSufxLen + strlen(acParDev) + 1;

			acSufx = new char[iSufxLen];
			acSufxDev = new char[iSufxDevLen];

			acSufxDev[0]='|';

			strncpy(&acSufx[0],acNewSuffix,iNewSufxLen);
			strncpy(&acSufxDev[1],acNewSuffix,iNewSufxLen);

			acSufx[0 + iNewSufxLen] = cNewChar;
			acSufxDev[1 + iNewSufxLen] = cNewChar;

			strcpy(&acSufx[1 + iNewSufxLen],acParSufx);
			strcpy(&acSufxDev[2 + iNewSufxLen],acParDev);

		} else {
			acSufx = "";
			acSufxDev = "|";
		}
		os << "[Suffix:" << acSufxDev << ",\"" << acSufx << "\"]";

		//display rule that applies to this node
		os << " ";
		ToString(os, iRuleAddress, 0, acSufx);

		//display hash table
		if ((iType & BitInternal) == BitInternal) {

			GETBYTEMOVE(, iSubsNum, ModLen);
			iSubs = new dword[iSubsNum];
			bSubs = new unsigned char[iSubsNum];

			ostringstream ossLine1;
			ostringstream ossLine2;
			ostringstream ossLine3;

			int iEmptyEntryNum = 0;
			for (int i = 0; i< iSubsNum; i++) {
				GETBYTEMOVE(int, cSubChar, CharLen);
				GETDWORDMOVE(dword, iSubAddres, AddrLen);
				bSubs[i]=cSubChar;
				iSubs[i]=iSubAddres;

				if (iSubAddres==0) {
					ossLine1 << " | NULL";
					ossLine2 << right << " |" << setw(5) << (int) i;
					ossLine3 << " | NULL";
					iEmptyEntryNum++;

				} else {
					ossLine1 << right << " |" << setw(3) << (char) cSubChar << "="  << setw(3) << (int) cSubChar;
					ossLine2 << right << " |" << setw(7) << (int) i;
					ossLine3 << right << " |" << setw(7) << iSubAddres;
				}
			}

			os << " HASHTABLE:";
			os << "[Size/Divider:" << iSubsNum << "]";
			os << "[Entries:" << iSubsNum - iEmptyEntryNum << "]";
			os << "[Unused:" << setprecision(4) << (float) 100 *iEmptyEntryNum/iSubsNum << "%]";
			os << " ";

			os << endl << setfill('\t') << setw(iDepth+1) << "" << "." << setfill('-') << setw(ossLine3.str().length() + 8) << ".";
			os << endl << setfill('\t') << setw(iDepth+1) << "" << "|  Pos:" << ossLine2.str() << " |";
			os << endl << setfill('\t') << setw(iDepth+1) << "" << "| Char:" << ossLine1.str() << " |";
			os << endl << setfill('\t') << setw(iDepth+1) << "" << "| Addr:" << ossLine3.str() << " |";
			os << endl << setfill('\t') << setw(iDepth+1) << "" << "'" << setfill('-') << setw(ossLine3.str().length() + 8) << "'";
			os << endl;
		}
	}

	//display sub nodes
	for (int i = 0; i<iSubsNum; i++)
		if (iSubs[i]!=NULL) {
			ToString(os, iSubs[i], iDepth + 1, acSufx, acSufxDev, bSubs[i]);
			if (i<iSubsNum-1) os << endl;
		}

	os.flush();
}


//-------------------------------------------------------------------------------------------
//returns nice valid c++ code representation memory block
void RdrLemmatizer::ToStringHex(ostream &os) const {
	os << dec << noshowbase;
	os << "#define RrdLemmData" << endl;
	os << "#define DATA_LEN " << iDataLen << endl;

	os << hex << right << setfill('0');
	os << "#define DATA_TBL {";

	int iLen = iDataLen/8;
	for(int i=0; i<iLen; i++) {
		if (i%5!=0) os << " ";
		else os << " \\" << endl << "\t";

		os << "0x" << setw(16) << ((qword*)abData)[i];

		if (i != iLen-1) os << ",";
	}

	os << " \\" << endl << "\t}" << endl;

	os.flush();
}

//-------------------------------------------------------------------------------------------
//loads all data needed from binary file
void RdrLemmatizer::LoadBinaryStream(istream &is) {
	iDataLen =0;
	is.read((char*) &iDataLen, 4);
	abData = new unsigned char[iDataLen];
	is.read((char*) abData, iDataLen);
}

//loads all data needed from binary file
void RdrLemmatizer::LoadBinary(const char *acFileName) {
	ifstream ifs(acFileName, ios_base::in | ios_base::binary);
	if(!ifs.good()){
		throw std::invalid_argument("Invalid model file or path.");
	}
	LoadBinaryStream(ifs);
	ifs.close();
}
